storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java

```diff
@@ -0,0 +1,458 @@
+                    execToComp, Time.currentTimeSecs(), "user");
+            topo.getComponents(); // sanity check - normally this should not fail
+
+            topoDetailsList.add(topo);
+        }
+        if (!errors.isEmpty() && failOnParseError) {
+            throw new Exception("Unable to parse all serialized objects\n\t" + String.join("\n\t", errors));
+        }
+        return topoDetailsList.toArray(new TopologyDetails[0]);
+    }
+
+    /**
+     * Check if the files in the resource directory are matched, can be read properly, and code/config files occur
+     * in matched pairs.
+     *
+     * @throws Exception showing bad and unmatched resource files.
+     */
+    @Test
+    public void testReadSerializedTopologiesAndConfigs() throws Exception {
+        List<String> resources = getResourceFiles(TEST_RESOURCE_PATH);
+        Assert.assertTrue("No resource files found in " + TEST_RESOURCE_PATH, !resources.isEmpty());
+        TopologyDetails[] topoDetailsArray = createTopoDetailsArray(true);
+    }
+
+    /**
+     * Create one supervisor and add to the supervisors list.
+     *
+     * @param rack rack-number
+     * @param superInRack supervisor number in the rack
+     * @param cpu percentage
+     * @param mem in megabytes
+     * @param numPorts number of ports on this supervisor
+     * @param sups returned map os supervisors
+     */
+    private static void createAndAddOneSupervisor(
+            int rack, int superInRack, double cpu, double mem, int numPorts,
+            Map<String, SupervisorDetails> sups) {
+
+        List<Number> ports = new LinkedList<>();
+        for (int p = 0; p < numPorts; p++) {
+            ports.add(p);
+        }
+        String superId = String.format("r%03ds%03d", rack, superInRack);
+        String hostId  = String.format("host-%03d-rack-%03d", superInRack, rack);
+        Map<String, Double> resourceMap = new HashMap<>();
+        resourceMap.put(Config.SUPERVISOR_CPU_CAPACITY, cpu);
+        resourceMap.put(Config.SUPERVISOR_MEMORY_CAPACITY_MB, mem);
+        resourceMap.put("network.resource.units", 50.0);
+        SupervisorDetails sup = new SupervisorDetails(superId,
+                hostId, null, ports,
+                NormalizedResources.RESOURCE_NAME_NORMALIZER.normalizedResourceMap(resourceMap));
+        sups.put(sup.getId(), sup);
+    }
+
+    /**
+     * Create supervisors.
+     *
+     * @param uniformSupervisors true if all supervisors are of the same size, false otherwise.
+     * @return supervisor details indexed by id
+     */
+    private static Map<String, SupervisorDetails> createSupervisors(boolean uniformSupervisors) {
+        Map<String, SupervisorDetails> retVal;
+        if (uniformSupervisors) {
+            int numRacks = 16;
+            int numSupersPerRack = 82;
+            int numPorts = 50; // scheduling will fill up 2-6, leaving of 90% workerslots unused - does this cause slow scheduling?
```

Comment:
What does the comment mean? Or can it be removed?

Reply:
I found that for topology components with large cpu/mem, most of the workeslots get left unused - but the looping logic in scheduling spends a lot of cycles in those. How workerslots are processed may need to be revisited for further speed gains. If it is not clear, it needs to be reworded.

Reply:
It would be better if it can be reworded. Your comment here is much clearer to me